File: src/main/java/frc/robot/subsystems/swerve/gyro/GyroIOSim.java
package frc.robot.subsystems.swerve.gyro;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import frc.robot.Robot;
import java.util.function.Supplier;

/** Simulated gyro that integrates chassis angular velocity. */
public final class GyroIOSim implements GyroIO {
  private final Supplier<ChassisSpeeds> speeds;

  public GyroIOSim(Supplier<ChassisSpeeds> speeds) {
    this.speeds = speeds;
  }

  @Override
  public void updateInputs(GyroIOInputs inputs) {
    double dt = Robot.getUpdateRateSec();
    double omega = speeds.get().omegaRadiansPerSecond;
    inputs.yaw = inputs.yaw.plus(Rotation2d.fromRadians(omega * dt));
    inputs.yawRateRadPerSec = omega;
    inputs.fullRotation = new Rotation3d(inputs.yaw);
    inputs.connected = true;
  }
}

File: src/main/java/frc/robot/subsystems/swerve/gyro/GyroIOSim.kt
package frc.robot.subsystems.swerve.gyro

import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.geometry.Rotation3d
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.units.Units.RadiansPerSecond
import java.util.function.Supplier

class GyroIOSim(
    private val speeds: Supplier<ChassisSpeeds>
) : GyroIO {
    override fun updateInputs(inputs: GyroIO.GyroInputs) {
        inputs.yaw += Rotation2d.fromRadians(speeds.get().omegaRadiansPerSecond * 0.02)
        inputs.yawRate.mut_replace(speeds.get().omegaRadiansPerSecond * 0.02, RadiansPerSecond)
        
        inputs.fullRotation = Rotation3d(inputs.yaw)
    }
}